ePBFT/src/pbft/libbyz/Replica.cpp

```diff
@@ -253,20 +253,15 @@ bool Replica::compare_execution_results(
   return true;
 }
 
-size_t Replica::ledger_cursor() const
-{
-  return ledger_replay->cursor();
-}
-
 bool Replica::apply_ledger_data(const std::vector<uint8_t>& data)
```

Comment:
Should this be called `apply_ledger_entry()` instead (and `process_data()` -> `process_entry()`). Data is somehow very generic and it's not clear if it is an arbitrary portion of the ledger or something specific.

Reply:
OK. Starring at this more, I reckon it'd be better if `Replica` only processed one entry at a time, instead of a chunk of the ledger. At the moment, the format of the framed ledger is known by `LedgerReplay.cpp` (in `serialized::skip(entry_data, data_size, sizeof(uint32_t));` line 37), which we want to avoid.

Reply:
I don't think I agree with this. We do the same in raft, we get a range of entries and then process them one by one, skipping the frame (`raft.h` and `ledgerenclave.h`). At some point, somewhere this will have to be done and `LedgerReplay` is as good place as any. I don't think we should restrict pbft to ask for the entries one-by-one if it is lagging behind.

Reply:
Good point! Doing the same thing that Raft does if best. I think we should make use of the existing `ledgerenclave.h` to hide the format of the framed ledger from `LedgerReplay.cpp`.

Reply:
I am not sure about using `ledgerenclave.h`. `LedgerEnclave` is used by raft to record entries (i.e. access the actual underlying ledger), and is also used by pbft to record entries to the ledger. However `LedgerReplay` does not want to go to that level, it is given some entries and it deserializes them without being concerned with the actual ledger. It can be seen as a helper method of `Replica` to actually reason about the entries and process them. In the raft equivalent it is what happens after `record_entry` returns with the data, and raft uses the data to modify the kv. The difference is that in raft we just need to put the data as it is received into the backup's ledger and then just apply the data to the kv, where as in pbft we need to reason about the data.

Reply:
After some discussion this will be addressed next: https://github.com/microsoft/CCF/issues/458